format/udf-spec.md

```diff
@@ -0,0 +1,285 @@
+| *required*  | `overload-id`              | `long`                                                 | Monotonically increasing identifier of this function overload.                                |
+| *required*  | `parameters`               | `list<parameter>`                                      | Ordered list of [function parameters](#parameter). Invocation order **must** match this list. |
+| *required*  | `return-type`              | `string` (Iceberg data type)                           | Return type. Example: `"string"`, `"struct<...>"`.                                            |
+| *required*  | `versions`                 | `list<overload-version>`                               | [Versioned implementations](#overload-version) of this overload.                              |
+| *required*  | `current-overload-version` | `long`                                                 | Monotonically increasing identifier of the current overload version.                          |
+| *optional*  | `function-type`            | `string` (either `"udf"` or `"udtf"`, default `"udf"`) | If `"udtf"`, `return-type` must be a `struct` describing the output schema.                   |
+| *optional*  | `doc`                      | `string`                                               | Documentation string.                                                                         |
+
+### Parameter
+| Requirement | Field  | Type                         | Description              |
+|-------------|--------|------------------------------|--------------------------|
+| *required*  | `name` | `string`                     | Parameter name.          |
+| *required*  | `type` | `string` (Iceberg data type) | Parameter data type.     |
+| *optional*  | `doc`  | `string`                     | Parameter documentation. |
+
+Notes:
+1. The `name` and `type` of a `parameter` are immutable. To change them, a new overload must be created. Only the optional documentation field (`doc`) can be updated in-place.
+2. The `return-type` is immutable. To change it, users must create a new overload and deprecate or remove the old one.
+
+### Overload-Version
+Each overload can evolve over time by introducing new versions. An overload version represents a specific implementation
+of the overload at a given point in time.
+
+| Requirement | Field name            | Type                                                                                 | Description                                                                             |
+|-------------|-----------------------|--------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------|
+| *required*  | `overload-version-id` | `long`                                                                               | Monotonically increasing identifier of the overload version.                            |
```

Comment:
How about `version-id` since we don't need more than one version concept in this spec?